backfiller/backfiller/main.py

```diff
@@ -0,0 +1,200 @@
+			raise
+		
+		local_segments = []
+
+	return local_segments
+
+def list_remote_hours(node, stream, variant, timeout=TIMEOUT):
+	
+	# just a wrapper around a call to restreamer.list_hours
+	uri = '{}/files/{}/{}'.format(node, stream, variant)
+	resp = requests.get(uri, timeout=timeout)
+	hours = resp.json()
+	return hours
+
+def list_remote_segments(node, stream, variant, hour, timeout=TIMEOUT):
+
+	# just a wrapper around a call to restreamer.list_segments
+	uri = '{}/files/{}/{}/{}'.format(node, stream, variant, hour_str)
+	resp = requests.get(uri, timeout=timeout)
+	remote_segments = resp.json()
+	return remote_segments
+
+# based on _get_segment in downloader/main
+# very basic error handling
+def get_remote_segment(base_dir, node, stream, variant, hour, missing_segment,
+			timeout=TIMEOUT):
+
+
+	path = os.path.join(base_dir, stream, variant, hour, missing_segment)
+	if os.path.exists(path):
+		return
+
+	substrs = path.split('-')
+	temp_path = '-'.join(substrs[:-1] + [str(uuid.uuid4()) + '.st'])
+	common.ensure_directory(temp_path)
+
+	uri = '{}/segments/{}/{}/{}/{}'.format(node, stream, variant, hour, missing_segment)
+	resp = requests.get(uri, stream=True, timeout=timeout)
+
+	with open(temp_path, 'w') as f:
+		for chunk in resp.iter_content(8192):
+			f.write(chunk)
+
+	common.rename(temp_path, path)
+
+
+
+def backfill(base_dir, stream, variants, hours=None, nodes=None):
+	
+	# loop over nodes backfilling from each
+
+	if nodes is None:
+		nodes = get_nodes()
+
+	#ideally do this in parallel
+	for node in nodes:
+		
+		try:
+			backfill_node(base_dir, node, stream, variants, hours)
+
+		#need to replace this with a more sophisticated error handler
+		except Exception as e:
+			print node, e
+
+
+def backfill_node(base_dir, node, stream, variants, hours=None, recent_cutoff=60):
+
+
+	# if hours is int, backfill last hours hourdirs
+	# else if hours is None, backfill all hourdirs
+	# else assume hours is iterable and backfill those hourdirs
+
+
+	if isinstance(hours, int):
+		n_hours = hours
+
+		if n_hours < 1:
+			raise ValueError('Number of hours has to be 1 or greater')
+
+		now = datetime.datetime.utcnow()
+		hours = [(now - i * timedelta(hours=1)).strftime(HOUR_FMT) for i in range(n_hours)]
+
+	for variant in variants:
+
+		if hours is None:
+			node_hours = list_remote_hours(node, stream, variant)
+		else:
+			node_hours = hours
+
+		for hour in node_hours:
+
+			local_segments = list_local_segments(base_dir, stream, variant, hour)
+			local_segments = set(local_segments)
+			remote_segments = list_remote_segments(node, stream, variant, hour)
+			remote_segments = set(remote_segments)
+			missing_segments = remote_segments - local_segments
+
+			for missing_segment in missing_segments:
+
+				#to avoid getting in the downloader's way ignore segments less than recent_cutoff old
+				time_str = '{}:{}'.format(hour, missing_segment.split('-')[0])
+				segment_time = datetime.datetime.strptime(time_str, HOUR_FMT + ':%M:%S.%f')
```

Comment:
huh. for some reason i had it in my head that datetime.strptime couldn't use `%f`, but i just tried it and it works. i should probably go back and eliminate unnessecary usage of `dateutil.parser.parse` (it's still needed for the HLS timestamps, which may have timezone information that needs to be corrected for before producing a timezone-less UTC datetime)